builtin/nomad/jobspec/releaser.go

```diff
@@ -0,0 +1,576 @@
+	jobClient := client.NomadClient.Jobs()
+	deploymentClient := client.NomadClient.Deployments()
+
+	st.Update("Getting job...")
+	jobs, _, err := jobClient.PrefixList(target.Name)
+	if err != nil {
+		return status.Errorf(codes.Aborted, "Unable to fetch Nomad jobs: %s", err.Error())
+	}
+
+	if len(jobs) > 0 {
+		if target.Name != jobs[0].ID {
+			st.Step(terminal.StatusError, fmt.Sprintf("Job could not be found, did you mean to promote %s?", jobs[0].ID))
+			return nil
+		}
+	} else {
+		status.Errorf(codes.Aborted, "Job not found: %s", err.Error())
+	}
+
+	q := &api.QueryOptions{
+		Namespace: jobs[0].JobSummary.Namespace,
+	}
+
+	job, _, err := jobClient.Info(jobs[0].ID, q)
+
+	if err != nil {
+		return status.Errorf(codes.Aborted, "Unable to fetch Nomad job: %s", err.Error())
+	}
+
+	st.Update("Getting latest deployments for job")
+	deploy, _, err := jobClient.LatestDeployment(*job.ID, q)
+	if err != nil {
+		return status.Errorf(codes.Aborted, "Unable to fetch latest deployment for Nomad job: %s", err.Error())
+	} else if deploy == nil {
+		st.Update("No active deployment for Nomad job.")
+		return nil
+	} else if deploy.JobVersion != *job.Version {
+		st.Update("Job version does not match deployment's job version.")
+		return nil
+	} else if deploy.Status != "running" {
+		st.Update("Deployment for job is no longer running.")
+		return nil
+	}
+
+	canaryDeployment := false
+	groupsToPromote := make([]string, len(deploy.TaskGroups))
+	for taskGroupName, taskGroup := range deploy.TaskGroups {
+		if r.config.Groups != nil {
+			if isElementExist(r.config.Groups, taskGroupName) && taskGroup.DesiredCanaries > 0 {
+				canaryDeployment = true
+				groupsToPromote = append(groupsToPromote, taskGroupName)
+				continue
+			}
+		} else if taskGroup.DesiredCanaries > 0 {
+			canaryDeployment = true
+			// If no groups to promote are specified in the config, promote all groups
+			//   that have canaries
+			groupsToPromote = append(groupsToPromote, taskGroupName)
+		}
+	}
+	if !canaryDeployment {
+		log.Info("Canaries not detected")
+		return nil
+	}
+
+	var currentTaskGroupState *api.DeploymentState
+	for _, group := range groupsToPromote {
+		if group != "" {
+			st.Update(fmt.Sprintf("Checking task group: %s", group))
+			// TODO: Update to pair a task group's name with its healthy deadline, so
+			//       we can set the deadline accordingly
+			//       d := time.Now().Add(time.Nanosecond * time.Duration(*job.TaskGroups[indexOfTaskGroupInSliceOfTaskGroupsOfJob].Update.HealthyDeadline))
+			d := time.Now().Add(time.Minute * time.Duration(5))
+			log.Debug(fmt.Sprintf("Healthy deadline: %s", d.String()))
+			ctx, cancel := context.WithDeadline(ctx, d)
+			defer cancel()
+			groupHealthy := false
+			for !groupHealthy {
+				currentTaskGroupState = deploy.TaskGroups[group]
+				if currentTaskGroupState.HealthyAllocs < len(currentTaskGroupState.PlacedCanaries) {
+					select {
+					case <-time.After(5 * time.Second): // forces sleep for 5 seconds
```

Comment:
@tgross thank you for the suggestion! I've updated the 5 second delay to instead use ticker to avoid the leak.

Also, I did not include another check for the deployment by calling the API at the top of the loop because earlier in the function [we get the deployment](https://github.com/hashicorp/waypoint/pull/2938/files/c80b5aca1fd787fa6f80e4a02715a82576e75c5a#diff-8465d427b10de27ab3337b90735cb0728c5f62a8ce0612476219fbb6aacc2574R111) for the first time. So unless the deployment status changed between that initial check and when we hit this loop, we still have a pretty current status of the deployment. I mostly did it this way to avoid a potentially likely unnecessary second call to the Nomad API, since we have a current deployment status already. However if you think making that second call is worth it to avoid hitting the select, I can definitely update it to get the deployment status again prior to hitting the select.